File: exercism/python/bracket-push/bracket_push.py

PAIRS = dict(zip('([{', ')]}'))


def check_brackets(xs):
    stack = []
    for x in xs:
        if x in PAIRS:
            stack.append(x)
        else:
            if len(stack) == 0 or not PAIRS[stack.pop()] == x:
                return False  # right skewed or mismatch respectively
    return len(stack) == 0  # ok or left skewed

File: exercism/javascript/bracket-push/bracket-push.js

const syms = {
  '(': ')',
  '{': '}',
  '[': ']'
}

const isBalanced = (str) => {
  const stack = []
  for (let char of str) {
    if (syms[char]) {
      stack.push(char)
      continue
    }
    const top = stack.pop()
    if (syms[top] !== char) return false  // mismatch or right-heavy
  }
  return stack.length === 0  // false if left-heavy
}

module.exports = isBalanced
